### DISCUSSION TITLE
Multiple domains and certificates

### DISCUSSION THREAD

**Discussion by frenkel** [basecamp/kamal#1100] [Q&A] [answered]
I have a number of domains that can get certificates through Let's Encrypt, a wildcard certificate from another Certificate Authority. My understanding is I won't be able to deploy this as one Kamal application. Is that correct? Or is there some trick to deploy the same appplication once, but with different certificates?

I think the following commands would need to be run by kamal:
```
kamal-proxy deploy service1 --target web-1:3000 --host something.com --tls
kamal-proxy deploy service2 --target web-1:3000 --host somethingelse.com --tls --tls-certificate-path cert.pem --tls-private-key-path key.pem
```

**Comment by rogerheykoop** [comment]
You can. 
Just try something like this in your proxy settings, but it will request a new certificate for you through letsencrypt. 

proxy:
  ssl: true
  hosts:
      - something.com
      - somethingelse.com

**Reply by frenkel** [reply]
Thanks, but for wildcard certificates that’s not possible. I nees the combination of with and without automatic certificates.

**Reply by yaKashif** [reply]
Having a wildcard host option by default is a good feature for tonnes of mult-tenant apps

```
proxy:
  ssl: true
  hosts:
    - something.com
    - somethingelse.com
    - *.somethingelse.com
```

**Reply by yaKashif** [reply]
Seems like https://github.com/basecamp/kamal-proxy/pull/45 works towards that but without auto ssl

**Reply by viktorianer** [reply]
And here in Kamal https://github.com/basecamp/kamal/pull/969

**Comment by frenkel** [comment]
I think this can be achieved by running `kamal-proxy deploy <new deployed container>` with the correct arguments for the other domains. Would it be possible to run this with a hook? post-deploy seems too late (small window where old container is missing), and pre-deploy will probably not have the new container available yet.

**Comment by frenkel** [comment, marked as answer]
I think this is a viable solution: https://github.com/basecamp/kamal/issues/1091#issuecomment-2498629542
It will generate a name that is kept between deploys so you can manually start kamal-proxy with different certificates/settings and point it to that name.

**Reply by viktorianer** [reply]
If it works great, but why I should start kamal-proxy manually?! Kamal is responsible to do this for me.

**Reply by frenkel** [reply]
Yes, you'll have to add some features to Kamal if you don't want to manually do this. I don't have the time and knowledge to add it, so this workaround is fine for me. It's a one time command I need to run on every server.

**Reply by frenkel** [reply]
By the way, if you want to make it part of you configuration/repo you can probably run it as a kamal hook.

**Reply by viktorianer** [reply]
> Yes, you'll have to add some features to Kamal if you don't want to manually do this. I don't have the time and knowledge to add it, so this workaround is fine for me. It's a one time command I need to run on every server.

Yes, me too. But, likely, people are working in this feature already 👍🏼. I linked it 3 weeks ago, just above your comment :).

**Reply by frenkel** [reply]
Yes, I saw. Hasn't been touched in 2 months so I started looking for workarounds. Thanks!

**Comment by adenta** [comment]
whats the latest on wildcard domains?

**Reply by viktorianer** [reply]
Follow
https://github.com/basecamp/kamal/pull/969

**Reply by adenta** [reply]
So from my skimming it’s _literally impossible_ to have a nice configuration for custom subdomains? You _need_ a custom, self managed cert?

**Reply by frenkel** [reply]
Yes, because the only way to get a wildcard certificate from Let's Encrypt is through DNS verification, which cannot be handled by Kamal/Kamal Proxy.

**Comment by tadast** [comment]
How would you configure a single box deployment where a wildcard domain is needed? I can set up letsencrypt with autoupdates on the server box, but there seems to be no option to configure a path to the cert files on the server.

Looks like https://github.com/basecamp/kamal/pull/1531 is great for when you have certs on the machine you're deploying from. But if I want to deploy a simple app from my dev machine I assume I need to employ hooks to sync the cert files from the server to my dev machine's kamal secrets, which feels a bit backwards when the cert files are already available on the server? Am I missing anything obvious?

**Reply by frenkel** [reply]
I think the idea is that bigger companies would store the certificates in a secret manager, which would be accessible from your deployment machine (locally or in some kind of CI). This would also scale to multiple servers as there is one central point where you replace the certificates when they expire.

You can use Bitwarden Secret Manager for example, which is free for small teams.

**Reply by tadast** [reply]
But that eliminates the point of automated certificate setup and renewals and makes you store secrets externally just so they can be sent back to the server (that's the source of truth!). 

Perhaps, if there is no solution for this at the moment, it should be a separate issue to see appetite for introducing extra config to support this

**Reply by frenkel** [reply]
You cannot automatically renew wildcard certificates with kamal-proxy. That needs a DNS mechanism which is not something kamal-proxy can offer. So if you want wildcard certificates you need to buy them somewhere and provide them to kamal on deploy.

**Reply by tadast** [reply]
The last sentence is incorrect, you can get free letsencrypt wildcard certificates that autorenew, and it's easy to set up with most popular DNS providers via certbot add-ons e.g. for AWS Route 53 you can install python3-certbot-dns-route53, configure credentials and then run `certbot certonly  --dns-route53  -d example.com  -d '*.example.com'  --agree-tos  -m you@example.com  --non-interactive` which will sort out the wildcart certificate and set up autorenewal directly on your server, so all kamal-proxy needs to do is grab certs from the file server, it just doesn't currently have an option to do that, or am I missing anything crucial here?

**Reply by frenkel** [reply]
You can only provide kamal-proxy with the required certificate on deploy. So if you want to do it with certbot it will need to do a new deploy for you after the certificate is updated.

**Reply by tadast** [reply]
Either that, or if a change is made to load certs from the filesystem, I suspect restarting the proxy container should be enough, in which case this could be triggered automatically after cert autoupdate?
